lldb/tools/lldb-dap/EventHelper.cpp

```diff
@@ -295,4 +307,267 @@ void SendMemoryEvent(DAP &dap, lldb::SBValue variable) {
+              dap_instance->modules.erase(module_id);
+              dap_instance->Send(protocol::Event{
+                  "module", protocol::ModuleEventBody{
+                                std::move(p_module).value(),
+                                protocol::ModuleEventBody::eReasonRemoved}});
+            } else if (module_exists) {
+              dap_instance->Send(protocol::Event{
+                  "module", protocol::ModuleEventBody{
+                                std::move(p_module).value(),
+                                protocol::ModuleEventBody::eReasonChanged}});
+            } else if (!remove_module) {
+              dap_instance->modules.insert(module_id);
+              dap_instance->Send(protocol::Event{
+                  "module", protocol::ModuleEventBody{
+                                std::move(p_module).value(),
+                                protocol::ModuleEventBody::eReasonNew}});
+            }
+          }
+        } else if (event_mask & lldb::SBTarget::eBroadcastBitNewTargetCreated) {
+          auto target = lldb::SBTarget::GetTargetFromEvent(event);
+
+          // Find the DAP instance that owns this target to check if we should
+          // ignore this event.
+          DAP *dap_instance = DAPSessionManager::FindDAP(target);
+
+          // Get the target and debugger IDs for the new session to use.
+          lldb::user_id_t target_id = target.GetGloballyUniqueID();
+          lldb::SBDebugger target_debugger = target.GetDebugger();
+          int debugger_id = target_debugger.GetID();
+
+          // We create an attach config that contains the debugger ID and target
+          // ID. The new DAP instance will use these IDs to find the existing
+          // debugger and target via FindDebuggerWithID and
+          // FindTargetByGloballyUniqueID.
+          llvm::json::Object attach_config;
+
+          attach_config.try_emplace("type", "lldb");
+          attach_config.try_emplace("debuggerId", debugger_id);
+          attach_config.try_emplace("targetId", target_id);
+          const char *session_name = target.GetTargetSessionName();
+          attach_config.try_emplace("name", session_name);
+
+          // 2. Construct the main 'startDebugging' request arguments.
+          llvm::json::Object start_debugging_args{
+              {"request", "attach"},
+              {"configuration", std::move(attach_config)}};
+
+          // Send the request. Note that this is a reverse request, so you don't
+          // expect a direct response in the same way as a client request.
+          // If we don't have a dap_instance (target wasn't found), get any
+          // active instance
+          if (!dap_instance) {
+            std::vector<DAP *> active_instances =
+                DAPSessionManager::GetInstance().GetActiveSessions();
+            if (!active_instances.empty())
+              dap_instance = active_instances[0];
+          }
+
+          if (dap_instance) {
```

Comment:
Should we report an error if we made it here without a dap_instance? That seems like an internal logic issue, right?